src/Raven.Client/Documents/AI/AiConversation.cs

```diff
@@ -107,7 +181,58 @@ public async Task<AiConversationResult> RunAsync(CancellationToken token = defau
             _userPrompt = null;
             _actionResponses.Clear();
         }
+    }
 
-        return _actionRequests.Count > 0 ? AiConversationResult.ActionRequired : AiConversationResult.Done;
+    private class AiActionContext<TActionParametersSchema>
+        where TActionParametersSchema : class
+    {
+        private readonly Func<TActionParametersSchema, Task<object>> _asyncAction;
+        private readonly AiHandleErrorStrategy _errorStrategy;
+        private readonly AiOperations _aiOperations;
+
+        public AiActionContext(AiOperations aiOperations, Func<TActionParametersSchema, Task<object>> asyncAction, AiHandleErrorStrategy errorStrategy)
+        {
+            _asyncAction = asyncAction;
+            _errorStrategy = errorStrategy;
+            _aiOperations = aiOperations;
+        }
+
+        public AiActionContext(AiOperations aiOperations, Func<TActionParametersSchema, object> asyncAction, AiHandleErrorStrategy errorStrategy)
+        {
+            _asyncAction = args => Task.FromResult(asyncAction(args));
+            _aiOperations = aiOperations;
+            _errorStrategy = errorStrategy;
+        }
+
+        public async Task<object> ExecuteAsync(JsonOperationContext context, string arguments, CancellationToken token = default)
+        {
+            if (typeof(TActionParametersSchema) == typeof(string))
+            {
+                var args = arguments as TActionParametersSchema;
+                return await Invoke(args).ConfigureAwait(false);
+            }
+
+            using (var json = context.Sync.ReadForMemory(arguments, "tool/arguments"))
```

Comment:
you need to use async. I know there is no impl for this, but you just need to copy this method and change the inner calles to async ones

Reply:
hmm, fine, this is using MemoryStream underneath. Can stay as is.